engine/collection/epochmgr/engine.go

```diff
@@ -108,25 +139,97 @@ func (e *Engine) createEpochComponents(epoch protocol.Epoch) (*EpochComponents,
 		return nil, fmt.Errorf("could not setup requirements for epoch (%d): %w", epoch, err)
 	}
 
-	reqs := &EpochComponents{
+	components := &EpochComponents{
 		state:    state,
 		prop:     prop,
 		sync:     sync,
 		hotstuff: hot,
 	}
-	return reqs, err
+	return components, err
+}
+
+// EpochTransition handles the epoch transition protocol event.
+func (e *Engine) EpochTransition(_ uint64, first *flow.Header) {
+	e.unit.Launch(func() {
+		err := e.onEpochTransition(first)
+		if err != nil {
+			// failing to complete epoch transition is a fatal error
+			e.log.Fatal().Err(err).Msg("failed to complete epoch transition")
+		}
+	})
 }
 
 // EpochSetupPhaseStarted handles the epoch setup phase started protocol event.
 func (e *Engine) EpochSetupPhaseStarted(_ uint64, _ *flow.Header) {
-	e.unit.Launch(e.prepareNextEpoch)
+	e.unit.Launch(e.onEpochSetupPhaseStarted)
+}
+
+// onEpochTransition is called when we transition to a new epoch. It arranges
+// to shut down the last epoch's components and starts up the new epoch's.
+func (e *Engine) onEpochTransition(first *flow.Header) error {
+	e.unit.Lock()
+	defer e.unit.Unlock()
+
+	epoch := e.state.Final().Epochs().Current()
+	counter, err := epoch.Counter()
+	if err != nil {
+		return fmt.Errorf("could not get epoch counter: %w", err)
+	}
+
+	log := e.log.With().
+		Uint64("epoch_height", first.Height).
+		Uint64("epoch_counter", counter).
+		Logger()
+
+	// exit early and log if the epoch already exists
+	_, exists := e.epochs[counter]
+	if exists {
+		log.Warn().Msg("epoch transition: components for new epoch already setup")
+		return nil
+	}
+
+	log.Info().Msg("epoch transition: creating components for new epoch...")
+
+	// create components for new epoch
+	components, err := e.createEpochComponents(epoch)
+	if err != nil {
+		return fmt.Errorf("could not create epoch components: %w", err)
+	}
+
+	// start up components
+	err = e.startEpochComponents(counter, components)
+	if err != nil {
+		return fmt.Errorf("could not start epoch components: %w", err)
+	}
+
+	log.Info().Msg("epoch transition: new epoch components started successfully")
+
+	// set up trigger to shut down previous epoch components after max expiry
+	e.heightEvents.OnHeight(first.Height+flow.DefaultTransactionExpiry, func() {
```

Comment:
This is a tricky problem. I'm not sure if the solution here would work well. This reminds me [the issue I ran into when implementing consensus tests](https://github.com/onflow/flow-go/blob/b6798cc07caac3809fe5b0d637a3ab54399ffa2d/consensus/integration/stopper_test.go#L27-L33). The issue is the faster nodes who have reached the targeted height will stop hotstuff, which will cause other slow nodes unable to get the blocks to finalize their targeted height, and those slow nodes will stuck forever.

In the consensus tests, I solve the issue by [letting faster nodes keep building blocks until all the honest nodes have all finalized enough blocks, and then stop them all](https://github.com/onflow/flow-go/blob/b6798cc07caac3809fe5b0d637a3ab54399ffa2d/consensus/integration/stopper_test.go#L55-L71). 

I feel we might want something similar here.

Reply:
I believe what you're describing doesn't apply here. Let me know if I'm misunderstanding.

We're dealing with two different chains here: the cluster chain and the main chain. We're triggering the end of the cluster chain based on reaching a particular height on the main chain. Progress on the main chain will continue independently of individual collectors shutting down their cluster chain, so all collectors will eventually shut down their cluster chain so long as they continue following the main chain.

Reply:
Sorry, I think you are right :) 

Glad this is not an issue.

Reply:
I would suggest to add some explanation on why the timing is chosen

Reply:
> Glad this is not an issue.

Me too 😀 -- I'll add some docs to explain what's going on here in more detail

Reply:
Added a note in https://github.com/onflow/flow-go/pull/2/commits/2b39f85451857ad343940b84db8f6c550af91d02